src/Cooked/MockChain/AutoReferenceScripts.hs

```diff
@@ -0,0 +1,64 @@
+-- | This module provides a function to ensure that each redeemer used in a
+-- skeleton is attached a reference scripts when a known utxos contains it.
+module Cooked.MockChain.AutoReferenceScripts (toTxSkelWithReferenceScripts) where
+
+import Control.Monad
+import Cooked.Conversion
+import Cooked.MockChain.BlockChain
+import Cooked.MockChain.UtxoSearch
+import Cooked.Output
+import Cooked.Skeleton
+import Data.Map qualified as Map
+import Data.Maybe
+import Optics.Core
+import PlutusLedgerApi.V3 qualified as Api
+
+-- | Searches through the known utxos for a utxo containing a reference script
+-- with a given script hash, and returns the first such utxo found, if any.
```

Comment:
Isn't there an optimization to be found such as setting a flag or adding an entry in a map of known ref scripts (script hash to tx out ref) in the chain state when a utxo with a ref script is produced? Could this improve performance by not going through everything and just looking up in the map?

Reply:
see response above